agent/structs/config_entry.go

```diff
@@ -688,6 +708,29 @@ type UpstreamConfig struct {
 	MeshGateway MeshGatewayConfig `json:",omitempty" alias:"mesh_gateway" `
 }
 
+func (cfg UpstreamConfig) Clone() UpstreamConfig {
+	cfg2 := cfg
+
+	cfg2.Limits = cfg.Limits.Clone()
+	cfg2.PassiveHealthCheck = cfg.PassiveHealthCheck.Clone()
+
+	return cfg2
+}
+
+func (cfg *UpstreamConfig) ServiceID() ServiceID {
```

Comment:
Is this used anywhere? Maybe we could just keep `ServiceName()`

Reply:
Yes, semantically an UpstreamConfig really should only discuss `ServiceName` things, since upstreams are to "all" instances instead of just _this one_, but `seenUpstreams` logic in the flattening RPC only operates on ServiceID which is why this is here.